src/rust/src/wallet.rs

```diff
@@ -244,6 +270,51 @@ impl Wallet {
     pub fn is_mine(&self, txid: &TxId) -> bool {
         matches!(self.wallet_txs.get(txid), Some(wtx) if !wtx.decrypted_notes.is_empty())
     }
+
+    pub fn get_filtered_notes(
+        &self,
+        ivk: &IncomingViewingKey,
+        ignore_spent: bool,
+        ignore_locked: bool,
+        require_spending_key: bool,
+    ) -> Vec<(OutPoint, DecryptedNote)> {
+        self.wallet_txs
+            .values()
+            .flat_map(|wallet_tx| {
+                wallet_tx
+                    .decrypted_notes
+                    .iter()
+                    .filter_map(move |(idx, dnote)| {
+                        let outpoint = OutPoint {
+                            txid: wallet_tx.txid,
+                            action_idx: *idx,
+                        };
+
+                        if (ignore_spent && self.spent_notes.contains(&outpoint))
+                            || (ignore_locked && self.locked_notes.contains(&outpoint))
+                            || (require_spending_key
+                                && self.key_store.spending_key_for_ivk(&dnote.ivk).is_none())
+                        {
+                            None
+                        } else if &dnote.ivk == ivk {
```

Comment:
If `ivk` is removed from `DecryptedNote` then these conditionals can be swapped around to simplify the code.